Examples and updated saving/loading

jonahdf opened the pull request:
The long awaited PR to master. 
- 100% passing tests
- Fully updated game state structs
- Integrated CLI files
- Hash table integration
- Selective saving/loading (only mutable structs)

Warning: this is a LARGE pull request, since we could not figure out how to merge smaller changes at a time (nearly all our code is dependent on the rest). This might take a while to review. There are also indeed merge conflicts that we must work out, but I figure it is better to get this beast started now and catch the problems in the reviewing. I have made @dbarer the reviewer from another team because he has helped integrate some CLI changes and has probably gotten more familiar with our code.

Also worth noting that top-level makefile does not compile as of now, but checkpointing, CLI, and common all do.

dbarer commented on src/checkpointing/include/load.h:
```diff
@@ -17,70 +19,79 @@
  * - -1 if unsuccessful
  */
-int load(char *filename, Game *g, game_t *g_t);
+int load(char *filename, game_t *g_t);
```
One thing I noticed is that you have the parameters switched around for save/load. It's int save(game_t *g_t, char *filename) and int load(char *filename, game_t *g_t). Obviously not major, but would increase consistency.

chrisfernandes18 replied:
^^

jonahdf replied:
We discussed this as a group, and we specified this way so the first parameter is input, and second is output. We will keep it the way it is.

dbarer commented on src/checkpointing/src/load.c:
```diff
@@ -5,198 +5,386 @@
+        fprintf(stderr, "no item id saved\n");
+    } else {
+        i_t->item_id = i->item_id;
     }
```
The way CLI and UI are integrated, printf style functions won't print to the UI due to the software used. If you want the user to see these messages, UI has a print_to_cli function in their print_functions.c file that you should call.

chrisfernandes18 commented on src/checkpointing/game.proto:
```diff
@@ -1,36 +1,83 @@
+    optional int32 attributes_len = 5; //Length of array (helpful for loading)
 }
 
+// message Condition {
```
Why is all of this commented out? Are we not saving condition? I get that we're not saving path but wouldn't we save condition since it is a mutable thing?

jonahdf replied:
Yup, not sure how that made it into dev

chrisfernandes18 commented on src/checkpointing/game.proto:
```diff
@@ -1,36 +1,83 @@
-    repeated Object objs = 5; //Array of all object ids in room
+    repeated Item items = 4; //Array of all items in room
+    optional int32 items_len = 5;
+    // Comment out path since we no longer store paths
```
Shouldn't we just delete this if we are not storing path in room anymore

chrisfernandes18 commented on src/checkpointing/include/save.h:
```diff
@@ -87,4 +103,5 @@ int write_to_file(char *filename, uint8_t *buffer, unsigned len);
  */
 int save(game_t *game, char *filename);
```
We should rename game to g_t to be consistent with load and then like David said we should swap the two around

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+    items[1]->long_desc = "wrinkled and probably needs to be washed";
+    dorm->items = items;
+
+    item_t** all_items = malloc(sizeof(item_t*) * 3);
```
Why do we malloc 3 spaces we only need 2 since there are only 2 items

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+    candy->item_id = "1234";
+    candy->short_desc = "salt water taffy";
+    candy->long_desc = "glues mouth shut";
+    all_items[2] = malloc(sizeof(item_t));
```
If we want to put the candy in there as well we should refactor it into the forloop

chrisfernandes18 replied:
or into the section before where we were putting things into the items array and then edit the for loop to be i < 3 and we can ignore the comment I put about malloc'ing 3 spots

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+        cr_assert_eq(itemsucc, 0,
+                     "load_item: loading item %d in load_room failed", i);
+    }
+    Item *candy = malloc(sizeof(Item));
```
Make a choice for the same thing I said in load room

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+    game__init(game);
+    
+    game->curr_room = "5";
+    game->curr_player = "chad";
```
Shouldn't we change this o incorporate game-state's function add curr player to game if they have it? I can't remember if that is an actual function but we should look into that and if there is one to set curr room to game or something along those lines.

jonahdf replied:
No, this is a .proto type, so we have to do it manually.

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+    player__init(player1);
+    player1->player_id = "chad";
+    player1->health = 1;
+    game->all_players = &player1;
```
Do we not add player to the hashtable of all players in game?

jonahdf replied:
They've changed it so there is now only 1 player

chrisfernandes18 commented on src/cli/include/operations.h:
```diff
@@ -92,6 +104,8 @@ char *save_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t *ctx);
  * - Said description as a string
  */
+
+
```
get rid of this white space?

chrisfernandes18 commented on src/cli/src/operations.c:
```diff
@@ -31,15 +30,30 @@ char *hist_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t *ctx)
+    return "Invalid Input, Save failed\n";
+  }
+  if (validate_filename(tokens[1]) == true){
+    int sv = save(ctx->game, tokens[1]);
```
When we decide which way to put it in our functions we might need to switch these two parameters around

chrisfernandes18 commented on src/cli/src/operations.c:
```diff
@@ -31,15 +30,30 @@ char *hist_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t *ctx)
+    return "Invalid Input, Load failed\n";
+  }
+  if (validate_filename(tokens[1]) == true){
+    int ld = load(tokens[1], ctx->game);
```
Look at the comment for save_operation

chrisfernandes18 commented on src/checkpointing/tests/test_load.c:
```diff
@@ -0,0 +1,250 @@
+
+    // Start creating our game
+
+    game_t *g_t = game_new("Welcome to my sex lair\n");
```
OMG we should change this lmaoo

dbarer commented:
The merge conflict listed will be fixed by a PR that CLI is currently pushing to master. However, I think we need to determine why we're getting undefined reference errors to save & load before pushing this to master. If you guys are free to meet today, maybe looking it over together will help. It's something with how we're linking the libraries together in the Makefile.

mpcsh commented on src/checkpointing/game.proto:
```diff
@@ -1,36 +1,92 @@
-    required string object_id = 1;
+//In Game-state implementation, attribute is a union. Proto doesn't support unions, so we use optional instead
+
+message Actiontype {
```
`Actiontype` -> `ActionType`. Make sure *all* the names in this file are changed this way.

mpcsh commented on src/checkpointing/src/load.c:
```diff
@@ -5,198 +5,388 @@
-	return -1;
+    if (i_t == NULL) {
+        fprintf(stderr, "given null item_t struct\n");
+        return -1;
```
Should use the common `SUCCESS` and `FAILURE` macros. Please make sure any explicit numeric return values in this file are changed this way.

mpcsh commented on src/checkpointing/src/load.c:
```diff
@@ -5,198 +5,388 @@
     return 0;
+}
+
+int count(game_t *g_t)
```
This method needs a more descriptive name - count *what*?

mpcsh commented on src/cli/Makefile:
```diff
@@ -1,8 +1,9 @@
 CC = gcc
-SRCS = src/cmd.c src/shell.c src/parser.c src/operations.c 
-CFLAGS = -fPIC -Wall -Wextra -O2 -g -I./include/ -I../common/include/ -I../game-state/include/ -I../action_management/include/ -I../ui/include/
+<<<<<<< HEAD
```
Uh oh. This shouldn't be here. Also, why are there changes present in the CLI makefile?

jonahdf replied:
Removed. There are changes because part of this merge is in integrating CLI. There will be more changes to other makefiles once checkpointing/make merges into dev. 

I asked you the other day whether we should modify other group makefiles so the build would pass in our PR to master, and you told me yes.

mpcsh commented on src/cli/src/cmd.c:
```diff
@@ -94,21 +94,24 @@ lookup_t **lookup_t_new()
-    add_entry("MAP", map_operation, t);
-    add_entry("SWITCH", switch_operation, t);
-    add_action_entries(t);
+ int lookup_t_init(lookup_t **t)
```
What happened here? Why is everything indented an extra space?

jonahdf replied:
Shouldn't we let the CLI style changes be fixed in their branch? We are just working on integration with save and load here.

mpcsh replied:
Yes, we should. That should mean no diff in this file

jonahdf replied:
fixed

mpcsh commented on src/cli/src/operations.c:
```diff
@@ -26,7 +27,29 @@ char *hist_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t *ctx)
 char *save_operation(char *tokens[TOKEN_LIST_SIZE], chiventure_ctx_t *ctx)
 {
-    return NULL;
+  if(tokens[1] == NULL){
```
Need spaces here. Please run everything through `astyle`.

jonahdf replied:
Same comment as above

mpcsh replied:
I don't think this is the same though - this looks like y'all added code to CLI

jonahdf replied:
fixed

mpcsh commented:
Also someone please merge master into this branch so that the build passes again

jonahdf commented:
> Also someone please merge master into this branch so that the build passes again

@mpcsh The build won't pass until we merge checkpointing/make. But I also did merge master

jonahdf commented:
We fixed the CLI formatting (or not formatting), updated the Makefiles (so travis finally passed with Borja's help), and responded to Christian and David's reviews (mostly variable naming/ordering). This also includes our demo branch, which updated some of our save and load features to avoid segmentation faults. We learned that this demo is not acceptable for our final presentation, but we kept it in this PR because it is a good example of all our save/load features. After Mark and Chris re-reviewed, we finally merged.

mpcsh commented:
Issue Score: ✓

Comments:
Well... This is admirable work, but we just can't give a ✓+ to something this massive. This should've been broken up and sent as 2 or 3 separate PRs.
<!-- {"score": "check"} -->
